src/mitol/payment_gateway/api.py

```diff
@@ -0,0 +1,206 @@
+"""
+API for the Payment Gateway
+"""
+import abc
+import hashlib
+import hmac
+import logging
+import uuid
+from base64 import b64encode
+
+from django.conf import settings
+
+from mitol.common.utils.datetime import now_in_utc
+from mitol.payment_gateway.constants import MITOL_PAYMENT_GATEWAY_CYBERSOURCE
+
+ISO_8601_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+
+class PaymentGateway(abc.ABC):
+    _GATEWAYS = {}
+
+    def __init_subclass__(cls, *, gateway_class, **kwargs):
+        super().__init_subclass__()
+
+        if gateway_class in cls._GATEWAYS:
+            raise TypeError(f"{gateway_class} has already been defined")
+
+        cls._GATEWAYS[gateway_class] = cls
+
+    @abc.abstractmethod
+    def prepare_checkout(self, order, cart, receipt_url, cancel_url, **kwargs):
+        """
+        This is the entrypoint to the payment gateway.
+        Args:
+            order: dict containing:
+                - username      Purchaser username
+                - ip_address    Purchaser's IP address
+                - reference     Order reference number
+            cart: list of dicts containing:
+                - code      Item code         (in xPro: content_type)
+                - name      Item name         (in xPro: description)
+                - quantity  Quantity          (in xPro: quantity)
+                - sku       SKU               (in xPro: content_object.id)
+                - taxable   Taxable amount    (in xPro: 0)
+                - unitprice Unit price        (in xPro: calculated price with discount)
```

Comment:
I'm thinking of a few improvements here:

- Remove `cart` and just nest a list of `items` under `order`
- Instead of `dict` objects, use dataclasses:
  - As implemented a `dict` would actually cause an error because the code in `CyberSourcePaymentGateway` accesses values by property instead of by key.
  - Dataclasses would provide the best of both worlds - both flexible data structure like a dict (via subclassing) and providing a typed interface. We probably won't use the extensibility of this quite yet, but it'll be good to have the right data type.